// @serenity-is/sleekgrid/dist/compat/plugins/slick.rowmovemanager.js
var Slick;
(Slick ||= {})._ = (() => {
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __export = (target, all) => {
    for (var name in all)
      __defProp(target, name, { get: all[name], enumerable: true });
  };
  var __copyProps = (to, from, except, desc) => {
    if (from && typeof from === "object" || typeof from === "function") {
      for (let key of __getOwnPropNames(from))
        if (!__hasOwnProp.call(to, key) && key !== except)
          __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
    }
    return to;
  };
  var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

  // src/plugins/rowmovemanager.ts
  var rowmovemanager_exports = {};
  __export(rowmovemanager_exports, {
    RowMoveManager: () => RowMoveManager
  });

  // ../domwise/dist/index.js
  var hasOwnProperty = Object.hasOwnProperty;
  var bindThisHandler = {
    get: (target, property) => {
      if (hasOwnProperty.call(target, property)) {
        return target[property];
      }
      const m = target[property];
      if (typeof m === "function") {
        return target[property] = m.bind(target);
      }
      return m;
    }
  };
  var bindThisProxyMap = /* @__PURE__ */ new WeakMap();
  function bindThis(obj) {
    let proxy = bindThisProxyMap.get(obj);
    if (!proxy) {
      proxy = new Proxy(obj, bindThisHandler);
      bindThisProxyMap.set(obj, proxy);
    }
    return proxy;
  }
  var Component = class {
    static {
      this.isComponent = true;
    }
    constructor(props) {
      this.props = props;
    }
    render() {
      return null;
    }
  };
  var i = /* @__PURE__ */ Symbol.for("preact-signals");
  function t() {
    if (!(s > 1)) {
      var i2, t2 = false;
      while (void 0 !== h) {
        var r2 = h;
        h = void 0;
        f++;
        while (void 0 !== r2) {
          var o2 = r2.o;
          r2.o = void 0;
          r2.f &= -3;
          if (!(8 & r2.f) && c(r2)) try {
            r2.c();
          } catch (r3) {
            if (!t2) {
              i2 = r3;
              t2 = true;
            }
          }
          r2 = o2;
        }
      }
      f = 0;
      s--;
      if (t2) throw i2;
    } else s--;
  }
  var o = void 0;
  function n(i2) {
    var t2 = o;
    o = void 0;
    try {
      return i2();
    } finally {
      o = t2;
    }
  }
  var h = void 0;
  var s = 0;
  var f = 0;
  var v = 0;
  function e(i2) {
    if (void 0 !== o) {
      var t2 = i2.n;
      if (void 0 === t2 || t2.t !== o) {
        t2 = { i: 0, S: i2, p: o.s, n: void 0, t: o, e: void 0, x: void 0, r: t2 };
        if (void 0 !== o.s) o.s.n = t2;
        o.s = t2;
        i2.n = t2;
        if (32 & o.f) i2.S(t2);
        return t2;
      } else if (-1 === t2.i) {
        t2.i = 0;
        if (void 0 !== t2.n) {
          t2.n.p = t2.p;
          if (void 0 !== t2.p) t2.p.n = t2.n;
          t2.p = o.s;
          t2.n = void 0;
          o.s.n = t2;
          o.s = t2;
        }
        return t2;
      }
    }
  }
  function u(i2, t2) {
    this.v = i2;
    this.i = 0;
    this.n = void 0;
    this.t = void 0;
    this.W = null == t2 ? void 0 : t2.watched;
    this.Z = null == t2 ? void 0 : t2.unwatched;
    this.name = null == t2 ? void 0 : t2.name;
  }
  u.prototype.brand = i;
  u.prototype.h = function() {
    return true;
  };
  u.prototype.S = function(i2) {
    var t2 = this, r2 = this.t;
    if (r2 !== i2 && void 0 === i2.e) {
      i2.x = r2;
      this.t = i2;
      if (void 0 !== r2) r2.e = i2;
      else n(function() {
        var i3;
        null == (i3 = t2.W) || i3.call(t2);
      });
    }
  };
  u.prototype.U = function(i2) {
    var t2 = this;
    if (void 0 !== this.t) {
      var r2 = i2.e, o2 = i2.x;
      if (void 0 !== r2) {
        r2.x = o2;
        i2.e = void 0;
      }
      if (void 0 !== o2) {
        o2.e = r2;
        i2.x = void 0;
      }
      if (i2 === this.t) {
        this.t = o2;
        if (void 0 === o2) n(function() {
          var i3;
          null == (i3 = t2.Z) || i3.call(t2);
        });
      }
    }
  };
  u.prototype.subscribe = function(i2) {
    var t2 = this;
    return E(function() {
      var r2 = t2.value, n2 = o;
      o = void 0;
      try {
        i2(r2);
      } finally {
        o = n2;
      }
    }, { name: "sub" });
  };
  u.prototype.valueOf = function() {
    return this.value;
  };
  u.prototype.toString = function() {
    return this.value + "";
  };
  u.prototype.toJSON = function() {
    return this.value;
  };
  u.prototype.peek = function() {
    var i2 = o;
    o = void 0;
    try {
      return this.value;
    } finally {
      o = i2;
    }
  };
  Object.defineProperty(u.prototype, "value", { get: function() {
    var i2 = e(this);
    if (void 0 !== i2) i2.i = this.i;
    return this.v;
  }, set: function(i2) {
    if (i2 !== this.v) {
      if (f > 100) throw new Error("Cycle detected");
      this.v = i2;
      this.i++;
      v++;
      s++;
      try {
        for (var r2 = this.t; void 0 !== r2; r2 = r2.x) r2.t.N();
      } finally {
        t();
      }
    }
  } });
  function c(i2) {
    for (var t2 = i2.s; void 0 !== t2; t2 = t2.n) if (t2.S.i !== t2.i || !t2.S.h() || t2.S.i !== t2.i) return true;
    return false;
  }
  function a(i2) {
    for (var t2 = i2.s; void 0 !== t2; t2 = t2.n) {
      var r2 = t2.S.n;
      if (void 0 !== r2) t2.r = r2;
      t2.S.n = t2;
      t2.i = -1;
      if (void 0 === t2.n) {
        i2.s = t2;
        break;
      }
    }
  }
  function l(i2) {
    var t2 = i2.s, r2 = void 0;
    while (void 0 !== t2) {
      var o2 = t2.p;
      if (-1 === t2.i) {
        t2.S.U(t2);
        if (void 0 !== o2) o2.n = t2.n;
        if (void 0 !== t2.n) t2.n.p = o2;
      } else r2 = t2;
      t2.S.n = t2.r;
      if (void 0 !== t2.r) t2.r = void 0;
      t2 = o2;
    }
    i2.s = r2;
  }
  function y(i2, t2) {
    u.call(this, void 0);
    this.x = i2;
    this.s = void 0;
    this.g = v - 1;
    this.f = 4;
    this.W = null == t2 ? void 0 : t2.watched;
    this.Z = null == t2 ? void 0 : t2.unwatched;
    this.name = null == t2 ? void 0 : t2.name;
  }
  y.prototype = new u();
  y.prototype.h = function() {
    this.f &= -3;
    if (1 & this.f) return false;
    if (32 == (36 & this.f)) return true;
    this.f &= -5;
    if (this.g === v) return true;
    this.g = v;
    this.f |= 1;
    if (this.i > 0 && !c(this)) {
      this.f &= -2;
      return true;
    }
    var i2 = o;
    try {
      a(this);
      o = this;
      var t2 = this.x();
      if (16 & this.f || this.v !== t2 || 0 === this.i) {
        this.v = t2;
        this.f &= -17;
        this.i++;
      }
    } catch (i3) {
      this.v = i3;
      this.f |= 16;
      this.i++;
    }
    o = i2;
    l(this);
    this.f &= -2;
    return true;
  };
  y.prototype.S = function(i2) {
    if (void 0 === this.t) {
      this.f |= 36;
      for (var t2 = this.s; void 0 !== t2; t2 = t2.n) t2.S.S(t2);
    }
    u.prototype.S.call(this, i2);
  };
  y.prototype.U = function(i2) {
    if (void 0 !== this.t) {
      u.prototype.U.call(this, i2);
      if (void 0 === this.t) {
        this.f &= -33;
        for (var t2 = this.s; void 0 !== t2; t2 = t2.n) t2.S.U(t2);
      }
    }
  };
  y.prototype.N = function() {
    if (!(2 & this.f)) {
      this.f |= 6;
      for (var i2 = this.t; void 0 !== i2; i2 = i2.x) i2.t.N();
    }
  };
  Object.defineProperty(y.prototype, "value", { get: function() {
    if (1 & this.f) throw new Error("Cycle detected");
    var i2 = e(this);
    this.h();
    if (void 0 !== i2) i2.i = this.i;
    if (16 & this.f) throw this.v;
    return this.v;
  } });
  function _(i2) {
    var r2 = i2.u;
    i2.u = void 0;
    if ("function" == typeof r2) {
      s++;
      var n2 = o;
      o = void 0;
      try {
        r2();
      } catch (t2) {
        i2.f &= -2;
        i2.f |= 8;
        b(i2);
        throw t2;
      } finally {
        o = n2;
        t();
      }
    }
  }
  function b(i2) {
    for (var t2 = i2.s; void 0 !== t2; t2 = t2.n) t2.S.U(t2);
    i2.x = void 0;
    i2.s = void 0;
    _(i2);
  }
  function g(i2) {
    if (o !== this) throw new Error("Out-of-order effect");
    l(this);
    o = i2;
    this.f &= -2;
    if (8 & this.f) b(this);
    t();
  }
  function p(i2, t2) {
    this.x = i2;
    this.u = void 0;
    this.s = void 0;
    this.o = void 0;
    this.f = 32;
    this.name = null == t2 ? void 0 : t2.name;
  }
  p.prototype.c = function() {
    var i2 = this.S();
    try {
      if (8 & this.f) return;
      if (void 0 === this.x) return;
      var t2 = this.x();
      if ("function" == typeof t2) this.u = t2;
    } finally {
      i2();
    }
  };
  p.prototype.S = function() {
    if (1 & this.f) throw new Error("Cycle detected");
    this.f |= 1;
    this.f &= -9;
    _(this);
    a(this);
    s++;
    var i2 = o;
    o = this;
    return g.bind(this, i2);
  };
  p.prototype.N = function() {
    if (!(2 & this.f)) {
      this.f |= 2;
      this.o = h;
      h = this;
    }
  };
  p.prototype.d = function() {
    this.f |= 8;
    if (!(1 & this.f)) b(this);
  };
  p.prototype.dispose = function() {
    this.d();
  };
  function E(i2, t2) {
    var r2 = new p(i2, t2);
    try {
      r2.c();
    } catch (i3) {
      r2.d();
      throw i3;
    }
    var o2 = r2.d.bind(r2);
    o2[Symbol.dispose] = o2;
    return o2;
  }

  // global-externals:_
  var { EventEmitter, EventSubscriber, H } = Slick;

  // src/plugins/rowmovemanager.ts
  var RowMoveManager = class _RowMoveManager {
    constructor(options) {
      this.handler = new EventSubscriber();
      this.onBeforeMoveRows = new EventEmitter();
      this.onMoveRows = new EventEmitter();
      this.options = Object.assign({}, _RowMoveManager.defaults, options);
    }
    static {
      this.defaults = {
        cancelEditOnDrag: false
      };
    }
    init(grid) {
      this.grid = grid;
      const boundThis = bindThis(this);
      this.handler.subscribe(grid.onDragInit, boundThis.handleDragInit).subscribe(grid.onDragStart, boundThis.handleDragStart).subscribe(grid.onDrag, boundThis.handleDrag).subscribe(grid.onDragEnd, boundThis.handleDragEnd);
    }
    destroy() {
      this.handler?.unsubscribeAll();
    }
    handleDragInit(e2) {
      e2.stopImmediatePropagation();
    }
    handleDragStart(e2, dd) {
      let cell = this.grid.getCellFromEvent(e2);
      if (this.options.cancelEditOnDrag && this.grid.getEditorLock().isActive()) {
        this.grid.getEditorLock().cancelCurrentEdit();
      }
      if (this.grid.getEditorLock().isActive() || !/move|selectAndMove/.test(this.grid.getColumns()[cell.cell].behavior)) {
        return false;
      }
      this.dragging = true;
      e2.stopImmediatePropagation();
      let selectedRows = this.grid.getSelectedRows();
      if (selectedRows.length == 0 || selectedRows.indexOf(cell.row) == -1) {
        selectedRows = [cell.row];
        this.grid.setSelectedRows(selectedRows);
      }
      let rowHeight = this.grid.getOptions().rowHeight;
      dd.selectedRows = selectedRows;
      let canvas = this.grid.getCanvasNode();
      const sp = dd.selectionProxy = canvas.appendChild(document.createElement("div"));
      sp.style = `position: absolute; z-index: 9999; width: ${canvas.clientWidth}px; height: ${rowHeight * selectedRows.length}px`;
      sp.className = "slick-row-move-proxy";
      const g2 = dd.guide = canvas.appendChild(document.createElement("div"));
      g2.style = `position: absolute; z-index: 9998; width: ${canvas.clientWidth}px; height: ${rowHeight}px; top: -1000px`;
      g2.className = "slick-row-move-guide";
      dd.insertBefore = -1;
    }
    handleDrag(e2, dd) {
      if (!this.dragging)
        return;
      e2.stopImmediatePropagation();
      let canvas = this.grid.getCanvasNode();
      let box = canvas.getBoundingClientRect();
      let docElem = document.documentElement;
      let canvasTop = box.top + window.scrollY - docElem.clientTop;
      let top = e2.pageY - canvasTop;
      dd.selectionProxy.style.top = top - 5 + "px";
      let insertBefore = Math.max(0, Math.min(Math.round(top / this.grid.getOptions().rowHeight), this.grid.getDataLength()));
      if (insertBefore !== dd.insertBefore) {
        let sgEvent = {
          rows: dd.selectedRows,
          insertBefore
        };
        if (this.onBeforeMoveRows.notify(sgEvent).getReturnValue() === false) {
          dd.guide.style.top = "-1000";
          dd.canMove = false;
        } else {
          dd.guide.style.top = insertBefore * this.grid.getOptions().rowHeight + "px";
          dd.canMove = true;
        }
        dd.insertBefore = insertBefore;
      }
    }
    handleDragEnd(e2, dd) {
      if (!this.dragging)
        return;
      this.dragging = false;
      e2.stopImmediatePropagation();
      dd.guide.remove();
      dd.selectionProxy.remove();
      if (dd.canMove) {
        let eventData = {
          rows: dd.selectedRows,
          insertBefore: dd.insertBefore
        };
        this.onMoveRows.notify(eventData);
      }
    }
  };
  return __toCommonJS(rowmovemanager_exports);
})();
["Data", "Editors", "Formatters", "Plugins"].forEach(ns => Slick._[ns] && (Slick[ns] = Object.assign(Slick[ns] || {}, Slick._[ns])) && delete Slick._[ns]); Object.assign(Slick, Slick._); delete Slick._;
