// @serenity-is/sleekgrid/dist/compat/slick.formatters.js
var Slick;
(Slick ||= {})._ = (() => {
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __export = (target, all) => {
    for (var name in all)
      __defProp(target, name, { get: all[name], enumerable: true });
  };
  var __copyProps = (to, from, except, desc) => {
    if (from && typeof from === "object" || typeof from === "function") {
      for (let key of __getOwnPropNames(from))
        if (!__hasOwnProp.call(to, key) && key !== except)
          __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
    }
    return to;
  };
  var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

  // src/formatters/index.ts
  var index_exports = {};
  __export(index_exports, {
    CheckBoxFormatter: () => CheckBoxFormatter,
    CheckmarkFormatter: () => CheckmarkFormatter,
    Formatters: () => Formatters,
    PercentCompleteBarFormatter: () => PercentCompleteBarFormatter,
    PercentCompleteFormatter: () => PercentCompleteFormatter,
    YesNoFormatter: () => YesNoFormatter
  });

  // global-externals:_
  var { escapeHtml, formatterContext } = Slick;

  // src/formatters/formatters.ts
  function PercentCompleteFormatter(ctx) {
    if (ctx.value == null || ctx.value === "")
      return "-";
    const span = document.createElement("span");
    span.textContent = ctx.value + "%";
    span.style.fontWeight = "bold";
    if (ctx.value < 50)
      span.style.color = "red";
    else
      span.style.color = "green";
    return span;
  }
  function PercentCompleteBarFormatter(ctx) {
    if (ctx.value == null || ctx.value === "")
      return "";
    var color;
    if (ctx.value < 30)
      color = "red";
    else if (ctx.value < 70)
      color = "silver";
    else
      color = "green";
    const span = document.createElement("span");
    span.className = "percent-complete-bar slick-percentcomplete-bar";
    span.style.background = color;
    span.style.width = ctx.value + "%";
    span.title = ctx.value + "%";
    return span;
  }
  function YesNoFormatter(ctx) {
    return ctx.value ? "Yes" : "No";
  }
  function CheckBoxFormatter(ctx) {
    const i = document.createElement("i");
    i.className = "slick-checkbox slick-edit-preclick" + (ctx.value ? " checked" : "");
    return i;
  }
  function CheckmarkFormatter(ctx) {
    if (!ctx.value)
      return "";
    const i = document.createElement("i");
    i.className = "slick-checkmark";
    return i;
  }

  // src/formatters/index.ts
  var Formatters;
  ((Formatters2) => {
    function PercentComplete(_row, _cell, value) {
      return PercentCompleteFormatter(formatterContext({ value }));
    }
    Formatters2.PercentComplete = PercentComplete;
    function PercentCompleteBar(_row, _cell, value) {
      return PercentCompleteBarFormatter(formatterContext({ value }));
    }
    Formatters2.PercentCompleteBar = PercentCompleteBar;
    function YesNo(_row, _cell, value) {
      return YesNoFormatter(formatterContext({ value }));
    }
    Formatters2.YesNo = YesNo;
    function Checkbox(_row, _cell, value) {
      return CheckBoxFormatter(formatterContext({ value }));
    }
    Formatters2.Checkbox = Checkbox;
    function Checkmark(_row, _cell, value) {
      return CheckmarkFormatter(formatterContext({ value }));
    }
    Formatters2.Checkmark = Checkmark;
  })(Formatters || (Formatters = {}));
  return __toCommonJS(index_exports);
})();
["Data", "Editors", "Formatters", "Plugins"].forEach(ns => Slick._[ns] && (Slick[ns] = Object.assign(Slick[ns] || {}, Slick._[ns])) && delete Slick._[ns]); Object.assign(Slick, Slick._); delete Slick._;
