// @serenity-is/sleekgrid/dist/compat/slick.core.js
var Slick;
(Slick ||= {})._ = (() => {
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __export = (target, all) => {
    for (var name in all)
      __defProp(target, name, { get: all[name], enumerable: true });
  };
  var __copyProps = (to, from, except, desc) => {
    if (from && typeof from === "object" || typeof from === "function") {
      for (let key of __getOwnPropNames(from))
        if (!__hasOwnProp.call(to, key) && key !== except)
          __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
    }
    return to;
  };
  var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

  // src/core/index.ts
  var index_exports = {};
  __export(index_exports, {
    CellRange: () => CellRange,
    EditorLock: () => EditorLock,
    EventDataWrapper: () => EventDataWrapper,
    EventEmitter: () => EventEmitter,
    EventSubscriber: () => EventSubscriber,
    GlobalEditorLock: () => GlobalEditorLock,
    Group: () => Group,
    GroupTotals: () => GroupTotals,
    NonDataRow: () => NonDataRow,
    addCssClass: () => addCssClass,
    applyFormatterResultToCellNode: () => applyFormatterResultToCellNode,
    basicDOMSanitizer: () => basicDOMSanitizer,
    columnDefaults: () => columnDefaults,
    convertCompatFormatter: () => convertCompatFormatter,
    defaultColumnFormat: () => defaultColumnFormat,
    disableSelection: () => disableSelection,
    escapeHtml: () => escapeHtml,
    formatterContext: () => formatterContext,
    gridDefaults: () => gridDefaults,
    initColumnProps: () => initColumnProps,
    keyCode: () => keyCode,
    parsePx: () => parsePx,
    preClickClassName: () => preClickClassName,
    removeCssClass: () => removeCssClass,
    titleize: () => titleize
  });

  // src/core/base.ts
  var NonDataRow = class {
    constructor() {
      this.__nonDataRow = true;
    }
  };
  var preClickClassName = "slick-edit-preclick";

  // src/core/cellrange.ts
  var CellRange = class {
    constructor(fromRow, fromCell, toRow, toCell) {
      if (toRow === void 0 && toCell === void 0) {
        toRow = fromRow;
        toCell = fromCell;
      }
      this.fromRow = Math.min(fromRow, toRow);
      this.fromCell = Math.min(fromCell, toCell);
      this.toRow = Math.max(fromRow, toRow);
      this.toCell = Math.max(fromCell, toCell);
    }
    /***
     * Returns whether a range represents a single row.
     */
    isSingleRow() {
      return this.fromRow == this.toRow;
    }
    /***
     * Returns whether a range represents a single cell.
     */
    isSingleCell() {
      return this.fromRow == this.toRow && this.fromCell == this.toCell;
    }
    /***
     * Returns whether a range contains a given cell.
     */
    contains(row, cell) {
      return row >= this.fromRow && row <= this.toRow && cell >= this.fromCell && cell <= this.toCell;
    }
    /***
     * Returns a readable representation of a range.
     */
    toString() {
      if (this.isSingleCell()) {
        return "(" + this.fromRow + ":" + this.fromCell + ")";
      } else {
        return "(" + this.fromRow + ":" + this.fromCell + " - " + this.toRow + ":" + this.toCell + ")";
      }
    }
  };

  // src/core/column.ts
  var columnDefaults = {
    resizable: true,
    sortable: false,
    minWidth: 30,
    rerenderOnResize: false,
    defaultSortAsc: true,
    focusable: true,
    selectable: true
  };
  function initColumnProps(columns, defaults) {
    var usedIds = {};
    for (var i = 0; i < columns.length; i++) {
      var m = columns[i];
      if (defaults != null) {
        for (var k in defaults) {
          if (m[k] === void 0)
            m[k] = defaults[k];
        }
      }
      if (m.minWidth && m.width < m.minWidth)
        m.width = m.minWidth;
      if (m.maxWidth && m.width > m.maxWidth)
        m.width = m.maxWidth;
      if (m.id == null || usedIds[m.id]) {
        const prefix = m.id != null && m.id.length ? m.id : m.field != null ? m.field : "col";
        var x = 0;
        while (usedIds[m.id = prefix + (x == 0 ? "" : "_" + x.toString())]) x++;
      }
      usedIds[m.id] = true;
      if (m.name === void 0) {
        m.name = titleize(m.field ?? m.id);
      }
    }
  }
  function titleize(str) {
    if (!str)
      return str;
    str = "" + str;
    let result = "";
    for (let i = 0; i < str.length; i++) {
      const char = str[i];
      const prevChar = i > 0 ? str[i - 1] : null;
      const nextChar = i < str.length - 1 ? str[i + 1] : null;
      if (char >= "A" && char <= "Z" && prevChar) {
        if (prevChar >= "a" && prevChar <= "z" || prevChar >= "0" && prevChar <= "9") {
          result += "_";
        } else if (prevChar >= "A" && prevChar <= "Z" && nextChar && nextChar >= "a" && nextChar <= "z") {
          let seqStart = i - 1;
          while (seqStart > 0 && str[seqStart - 1] >= "A" && str[seqStart - 1] <= "Z") {
            seqStart--;
          }
          if (i > seqStart) {
            result += "_";
          }
        }
      }
      if (char === "-" || /\s/.test(char)) {
        result += "_";
      } else {
        result += char;
      }
    }
    result = result.toLowerCase().replace(/_+/g, "_");
    return result.split("_").filter((x) => x.length).map((x) => x.charAt(0).toUpperCase() + x.substring(1).toLowerCase()).join(" ");
  }

  // src/core/editing.ts
  var EditorLock = class {
    /***
     * Returns true if a specified edit controller is active (has the edit lock).
     * If the parameter is not specified, returns true if any edit controller is active.
     * @param editController {EditController}
     * @return {Boolean}
     */
    isActive(editController) {
      return editController ? this.activeEditController === editController : this.activeEditController != null;
    }
    /***
     * Sets the specified edit controller as the active edit controller (acquire edit lock).
     * If another edit controller is already active, and exception will be thrown.
     * @param editController {EditController} edit controller acquiring the lock
     */
    activate(editController) {
      if (editController === this.activeEditController) {
        return;
      }
      if (this.activeEditController != null) {
        throw "SleekGrid.EditorLock.activate: an editController is still active, can't activate another editController";
      }
      if (!editController.commitCurrentEdit) {
        throw "SleekGrid.EditorLock.activate: editController must implement .commitCurrentEdit()";
      }
      if (!editController.cancelCurrentEdit) {
        throw "SleekGrid.EditorLock.activate: editController must implement .cancelCurrentEdit()";
      }
      this.activeEditController = editController;
    }
    /***
     * Unsets the specified edit controller as the active edit controller (release edit lock).
     * If the specified edit controller is not the active one, an exception will be thrown.
     * @param editController {EditController} edit controller releasing the lock
     */
    deactivate(editController) {
      if (this.activeEditController !== editController) {
        throw "SleekGrid.EditorLock.deactivate: specified editController is not the currently active one";
      }
      this.activeEditController = null;
    }
    /***
     * Attempts to commit the current edit by calling "commitCurrentEdit" method on the active edit
     * controller and returns whether the commit attempt was successful (commit may fail due to validation
     * errors, etc.).  Edit controller's "commitCurrentEdit" must return true if the commit has succeeded
     * and false otherwise.  If no edit controller is active, returns true.
     * @return {Boolean}
     */
    commitCurrentEdit() {
      return this.activeEditController ? this.activeEditController.commitCurrentEdit() : true;
    }
    /***
     * Attempts to cancel the current edit by calling "cancelCurrentEdit" method on the active edit
     * controller and returns whether the edit was successfully cancelled.  If no edit controller is
     * active, returns true.
     * @return {Boolean}
     */
    cancelCurrentEdit() {
      return this.activeEditController ? this.activeEditController.cancelCurrentEdit() : true;
    }
  };
  var GlobalEditorLock = new EditorLock();

  // src/core/event.ts
  var eventDataInitialized = false;
  function addEventDataProp(name, isArgs) {
    Object.defineProperty(EventDataWrapper.prototype, name, {
      enumerable: true,
      configurable: true,
      get: function() {
        if (isArgs) {
          if (typeof this.args === "object" && this.args !== null)
            return this.args[name];
        } else if (this.nativeEvent) {
          return this.nativeEvent[name];
        }
      },
      set: function(value) {
        if (isArgs) {
          if (typeof this.args === "object" && this.args !== null) {
            this.args[name] = value;
            return;
          }
        } else if (this.nativeEvent) {
          this.nativeEvent[name] = value;
          return;
        }
        Object.defineProperty(this, name, {
          enumerable: true,
          configurable: true,
          writable: true,
          value
        });
      }
    });
  }
  function initializeEventDataProps() {
    for (const key of [
      "altKey",
      "char",
      "bubbles",
      "button",
      "buttons",
      "cancelable",
      "changedTouches",
      "charCode",
      "clientX",
      "clientY",
      "code",
      "ctrlKey",
      "currentTarget",
      "detail",
      "eventPhase",
      "key",
      "keyCode",
      "metaKey",
      "offsetX",
      "offsetY",
      "originalEvent",
      "pageX",
      "pageY",
      "pointerId",
      "pointerType",
      "screenX",
      "screenY",
      "shiftKey",
      "relatedTarget",
      "target",
      "targetTouches",
      "toElement",
      "touches",
      "type",
      "view",
      "which"
    ]) {
      addEventDataProp(key, false);
    }
    for (const key of ["grid", "column", "node", "row", "cell", "item"]) {
      addEventDataProp(key, true);
    }
  }
  var EventDataWrapper = class {
    constructor(event, args) {
      this._isPropagationStopped = false;
      this._isImmediatePropagationStopped = false;
      this._isDefaultPrevented = false;
      this._returnValues = [];
      this._nativeEvent = event;
      this._args = args || {};
      if (!eventDataInitialized) {
        eventDataInitialized = true;
        initializeEventDataProps();
      }
    }
    get defaultPrevented() {
      return this._isDefaultPrevented;
    }
    preventDefault() {
      this._isDefaultPrevented = true;
      this._nativeEvent?.preventDefault?.();
    }
    isDefaultPrevented() {
      if (this._isDefaultPrevented)
        return true;
      if (this._nativeEvent && "defaultPrevented" in this._nativeEvent)
        return !!this._nativeEvent.defaultPrevented;
      if (this._nativeEvent && typeof this._nativeEvent.isDefaultPrevented === "function")
        return this._nativeEvent.isDefaultPrevented();
      return false;
    }
    /***
     * Stops event from propagating up the DOM tree.
     */
    stopPropagation() {
      this._isPropagationStopped = true;
      this._nativeEvent?.stopPropagation();
    }
    /***
     * Returns whether stopPropagation was called on this event object.
     */
    isPropagationStopped() {
      return this._isPropagationStopped;
    }
    /***
     * Prevents the rest of the handlers from being executed.
     */
    stopImmediatePropagation() {
      this._isImmediatePropagationStopped = true;
      this._nativeEvent?.stopImmediatePropagation();
    }
    /***
     * Returns whether stopImmediatePropagation was called on this event object.\
     */
    isImmediatePropagationStopped() {
      return this._isImmediatePropagationStopped;
    }
    get args() {
      return this._args;
    }
    addReturnValue(value) {
      this._returnValues.push(value);
      if (value !== void 0)
        this._returnValue = value;
    }
    getReturnValues() {
      return this._returnValues;
    }
    getReturnValue() {
      return this._returnValue;
    }
    get nativeEvent() {
      return this._nativeEvent;
    }
  };
  var EventEmitter = class {
    constructor() {
      this._handlers = [];
    }
    /***
     * Adds an event handler to be called when the event is fired.
     * <p>Event handler will receive two arguments - an <code>EventData</code> and the <code>data</code>
     * object the event was fired with.<p>
     * @param fn {Function} Event handler.
     */
    subscribe(fn) {
      this._handlers.push(fn);
    }
    /***
     * Removes an event handler added with <code>subscribe(fn)</code>.
     * @param fn {Function} Event handler to be removed.
     */
    unsubscribe(fn) {
      for (var i = this._handlers.length - 1; i >= 0; i--) {
        if (this._handlers[i] === fn) {
          this._handlers.splice(i, 1);
        }
      }
    }
    /***
     * Fires an event notifying all subscribers.
     * @param args {Object} Additional data object to be passed to all handlers.
     * @param e {EventDataWrapper}
     *      Optional.
     *      An <code>EventData</code> object to be passed to all handlers.
     *      For DOM events, an existing W3C/jQuery event object can be passed in.
     * @param scope {Object}
     *      Optional.
     *      The scope ("this") within which the handler will be executed.
     *      If not specified, the scope will be set to the <code>Event</code> instance.
     */
    notify(args, e, scope) {
      const sed = new EventDataWrapper(e, args);
      scope = scope || this;
      for (var i = 0; i < this._handlers.length && !(sed.isPropagationStopped() || sed.isImmediatePropagationStopped()); i++) {
        const returnValue = this._handlers[i].call(scope, sed, args);
        sed.addReturnValue(returnValue);
      }
      return sed;
    }
    clear() {
      this._handlers = [];
    }
  };
  var EventSubscriber = class {
    constructor() {
      this._handlers = [];
    }
    subscribe(event, handler) {
      this._handlers.push({
        event,
        handler
      });
      event.subscribe(handler);
      return this;
    }
    unsubscribe(event, handler) {
      var i = this._handlers.length;
      while (i--) {
        if (this._handlers[i].event === event && this._handlers[i].handler === handler) {
          this._handlers.splice(i, 1);
          event.unsubscribe(handler);
          return this;
        }
      }
      return this;
    }
    unsubscribeAll() {
      var i = this._handlers.length;
      while (i--) {
        this._handlers[i].event.unsubscribe(this._handlers[i].handler);
      }
      this._handlers = [];
      return this;
    }
  };
  var keyCode = {
    BACKSPACE: 8,
    DELETE: 46,
    DOWN: 40,
    END: 35,
    ENTER: 13,
    ESCAPE: 27,
    HOME: 36,
    INSERT: 45,
    LEFT: 37,
    PAGEDOWN: 34,
    PAGEUP: 33,
    RIGHT: 39,
    TAB: 9,
    UP: 38
  };

  // src/core/gridoptions.ts
  var gridDefaults = {
    addNewRowCssClass: "new-row",
    alwaysAllowHorizontalScroll: false,
    alwaysShowVerticalScroll: false,
    asyncEditorLoadDelay: 100,
    asyncEditorLoading: false,
    asyncPostCleanupDelay: 40,
    asyncPostRenderDelay: -1,
    autoEdit: true,
    autoHeight: false,
    cellFlashingCssClass: "flashing",
    dataItemColumnValueExtractor: null,
    defaultColumnWidth: 80,
    defaultFormat: defaultColumnFormat,
    editable: false,
    editorFactory: null,
    editorLock: GlobalEditorLock,
    enableAddRow: false,
    enableAsyncPostRender: false,
    enableAsyncPostRenderCleanup: false,
    enableCellNavigation: true,
    enableColumnReorder: true,
    enableHtmlRendering: false,
    enableTabKeyNavigation: true,
    enableTextSelectionOnCells: false,
    explicitInitialization: false,
    footerRowHeight: null,
    forceFitColumns: false,
    forceSyncScrolling: false,
    forceSyncScrollInterval: 250,
    formatterFactory: null,
    fullWidthRows: false,
    groupingPanel: false,
    groupingPanelHeight: null,
    headerRowHeight: null,
    leaveSpaceForNewRows: false,
    minBuffer: 3,
    multiColumnSort: false,
    multiSelect: true,
    renderAllCells: false,
    rowHeight: 30,
    selectedCellCssClass: "selected",
    showCellSelection: true,
    showColumnHeader: true,
    showFooterRow: false,
    showGroupingPanel: true,
    showHeaderRow: false,
    showTopPanel: false,
    suppressActiveCellChangeOnEdit: false,
    topPanelHeight: null,
    useCssVars: true
  };

  // src/core/util.tsx
  function addCssClass(el, cls) {
    if (cls == null || !cls.length)
      return;
    if (cls.indexOf(" ") >= 0) {
      var arr = cls.split(" ").map((x) => x.trim()).filter((x) => x.length);
      for (var a of arr)
        el.classList.add(a);
    } else
      el.classList.add(cls);
  }
  var esc = {
    "<": "&lt;",
    ">": "&gt;",
    '"': "&quot;",
    "'": "&apos;",
    "&": "&amp;"
  };
  function escFunc(a) {
    return esc[a];
  }
  function escapeHtml(s) {
    if (!arguments.length && this && this !== globalThis) {
      s = this.value;
    }
    if (s == null)
      return "";
    if (typeof s !== "string")
      s = "" + s;
    if (this && this !== globalThis && this.enableHtmlRendering === false)
      return s;
    return s.replace(/[<>"'&]/g, escFunc);
  }
  var maybeHtmlRegex = /<|>|&|"|'/;
  function basicDOMSanitizer(dirtyHtml) {
    if (!dirtyHtml) {
      return "";
    }
    if (!maybeHtmlRegex.test(dirtyHtml)) {
      return dirtyHtml;
    }
    if (typeof DOMParser === "undefined") {
      return escapeHtml(dirtyHtml);
    }
    const SAFE_URL_PATTERN = /^(?!javascript:|data:|vbscript:)(?:[a-z0-9+.-]+:|[^&:/?#]*(?:[/?#]|$))/i;
    try {
      const parser = new DOMParser();
      const doc = parser.parseFromString(dirtyHtml, "text/html");
      const body = doc.body || (typeof document !== "undefined" ? document.createElement("body") : null);
      if (!body) {
        return escapeHtml(dirtyHtml);
      }
      let targetElement = body;
      if (body.innerHTML.trim() === "" && doc.documentElement && doc.documentElement.innerHTML.trim()) {
        body.innerHTML = doc.documentElement.innerHTML;
      }
      if (body.innerHTML.trim() === "" && doc.documentElement) {
        targetElement = doc.documentElement;
      }
      const dangerousElements = targetElement.querySelectorAll("script, iframe, object, embed, form, input, button, textarea, select, style, link");
      dangerousElements.forEach((el) => el.remove());
      const allElements = targetElement.querySelectorAll("*");
      allElements.forEach((el) => {
        const element = el;
        Array.from(element.attributes).forEach((attr) => {
          const name = attr.name.toLowerCase();
          const value = attr.value;
          if (name.startsWith("on")) {
            element.removeAttribute(attr.name);
            return;
          }
          if (["href", "src", "xlink:href"].indexOf(name) >= 0) {
            if (!SAFE_URL_PATTERN.test(value)) {
              element.removeAttribute(attr.name);
              return;
            }
          }
          if (value.toLowerCase().indexOf("javascript") >= 0) {
            element.removeAttribute(attr.name);
          }
        });
      });
      return targetElement.innerHTML;
    } catch (e) {
      return escapeHtml(dirtyHtml);
    }
  }
  function disableSelection(target) {
    if (target) {
      target.style.userSelect = "none";
      target.addEventListener("selectstart", () => false);
    }
  }
  function removeCssClass(el, cls) {
    if (cls == null || !cls.length)
      return;
    if (cls.indexOf(" ") >= 0) {
      var arr = cls.split(" ").map((x) => x.trim()).filter((x) => x.length);
      for (var a of arr)
        el.classList.remove(a);
    } else
      el.classList.remove(cls);
  }
  function parsePx(str) {
    var value = parseFloat(str);
    if (isNaN(value))
      return 0;
    return value;
  }

  // src/core/formatting.ts
  function defaultColumnFormat(ctx) {
    if (ctx?.escape)
      return ctx.escape();
    if (!ctx?.enableHtmlRendering) {
      if (ctx.value == null)
        return "";
      if (typeof ctx.value !== "string")
        return "" + ctx.value;
      return ctx.value;
    }
    return escapeHtml(ctx?.value);
  }
  function convertCompatFormatter(compatFormatter) {
    if (compatFormatter == null)
      return null;
    return function(ctx) {
      var fmtResult = compatFormatter(ctx.row, ctx.cell, ctx.value, ctx.column, ctx.item, ctx.grid);
      if (fmtResult != null && typeof fmtResult !== "string" && Object.prototype.toString.call(fmtResult) === "[object Object]") {
        ctx.addClass = fmtResult.addClasses;
        ctx.tooltip = fmtResult.toolTip;
        return fmtResult.text;
      }
      return fmtResult;
    };
  }
  function applyFormatterResultToCellNode(ctx, fmtResult, node, opt) {
    if (!opt?.contentOnly) {
      var oldFmtAtt = node.dataset.fmtatt;
      if (oldFmtAtt?.length > 0) {
        for (var k of oldFmtAtt.split(","))
          node.removeAttribute(k);
        delete node.dataset.fmtatt;
      }
      var oldFmtCls = node.dataset.fmtcls;
      if (oldFmtCls?.length && ctx.addClass != oldFmtCls) {
        removeCssClass(node, oldFmtCls);
        if (!ctx.addClass?.length)
          delete node.dataset.fmtcls;
      }
      var oldTooltip = node.getAttribute("tooltip");
      if (oldTooltip != null && ctx.tooltip != oldTooltip)
        node.removeAttribute("tooltip");
      if (ctx.tooltip !== void 0 && oldTooltip != ctx.tooltip)
        node.setAttribute("tooltip", ctx.tooltip);
    }
    if (fmtResult == void 0)
      node.innerHTML = "";
    else if (fmtResult instanceof Node) {
      node.appendChild(fmtResult);
    } else if (!ctx.enableHtmlRendering)
      node.innerHTML = (ctx.sanitizer ?? escapeHtml)("" + fmtResult);
    else
      node.textContent = "" + fmtResult;
    if (!opt?.contentOnly) {
      if (ctx.addAttrs != null) {
        var keys = Object.keys(ctx.addAttrs);
        if (keys.length) {
          for (var k of keys) {
            node.setAttribute(k, ctx.addAttrs[k]);
          }
          node.dataset.fmtatt = keys.join(",");
        }
      }
      if (ctx.addClass?.length) {
        addCssClass(node, ctx.addClass);
        node.dataset.fmtcls = ctx.addClass;
      }
    }
  }
  function formatterContext(opt) {
    const gridOptions = opt?.grid?.getOptions?.();
    return {
      ...opt,
      enableHtmlRendering: opt?.enableHtmlRendering ?? gridOptions?.enableHtmlRendering ?? gridDefaults.enableHtmlRendering ?? false,
      escape: opt?.escape ?? escapeHtml,
      sanitizer: opt?.sanitizer ?? gridOptions?.sanitizer ?? gridDefaults.sanitizer ?? // @ts-ignore
      (typeof DOMPurify !== "undefined" && typeof DOMPurify.sanitize == "function" ? DOMPurify.sanitize : basicDOMSanitizer),
      value: opt?.value
    };
  }

  // src/core/group.ts
  var Group = class extends NonDataRow {
    constructor() {
      super(...arguments);
      this.__group = true;
      /**
       * Grouping level, starting with 0.
       * @property level
       * @type {Number}
       */
      this.level = 0;
      /***
       * Number of rows in the group.
       * @property count
       * @type {Number}
       */
      this.count = 0;
      /***
       * Whether a group is collapsed.
       * @property collapsed
       * @type {Boolean}
       */
      this.collapsed = false;
      /**
       * Rows that are part of the group.
       * @property rows
       * @type {Array}
       */
      this.rows = [];
    }
    /***
     * Compares two Group instances.
     * @return {Boolean}
     * @param group {Group} Group instance to compare to.
     */
    equals(group) {
      return this.value === group.value && this.count === group.count && this.collapsed === group.collapsed;
    }
  };
  var GroupTotals = class extends NonDataRow {
    constructor() {
      super(...arguments);
      this.__groupTotals = true;
      /***
       * Whether the totals have been fully initialized / calculated.
       * Will be set to false for lazy-calculated group totals.
       * @param initialized
       * @type {Boolean}
       */
      this.initialized = false;
    }
  };
  return __toCommonJS(index_exports);
})();
["Data", "Editors", "Formatters", "Plugins"].forEach(ns => Slick._[ns] && (Slick[ns] = Object.assign(Slick[ns] || {}, Slick._[ns])) && delete Slick._[ns]); Object.assign(Slick, Slick._); delete Slick._; Slick.Event = Slick.EventEmitter; Slick.EventHandler = Slick.EventSubscriber; Slick.Range = Slick.CellRange; typeof Map !== 'undefined' && (Slick.Map = Map);
